### PR TITLE
[web] Add support for ColorFilter on images

### PR DESCRIPTION
**Pull request by ferhatb**
Related issues: https://github.com/flutter/flutter/issues/56071 , https://github.com/flutter/flutter/issues/50897

Verified by comparing native vs web. Golden tests results: https://github.com/flutter/goldens/pull/92

Screen Shot 2020-05-03 at 11 17 59 AM

### CODE REVIEW

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 1]
```diff
@@ -797,3 +971,110 @@ String _maskFilterToCss(ui.MaskFilter maskFilter) {
...
+// A' = a1*R + a2*G + a3*B + a4*A + a5
+String _srcInColorFilterToSvg(ui.Color color) {
+  _filterIdCounter += 1;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" '
+      'filterUnits="objectBoundingBox" x="0%" y="0%" width="100%" height="100%">'
+      '<feColorMatrix values="0 0 0 0 1 ' // Ignore input, set it to absolute.
+          '0 0 0 0 1 '
+          '0 0 0 0 1 '
+          '0 0 0 1 0" result="destalpha"/>' // Just take alpha channel of of destination
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/bitmap_canvas.dart:998]:
```suggestion
          '0 0 0 1 0" result="destalpha"/>' // Just take alpha channel of destination
```

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 2]
```diff
@@ -797,3 +971,110 @@ String _maskFilterToCss(ui.MaskFilter maskFilter) {
...
+  _filterIdCounter += 1;
+  return '<svg width="0" height="0">'
+      '<filter id="_fcf$_filterIdCounter" filterUnits="objectBoundingBox" '
+      'x="0%" y="0%" width="100%" height="100%">'
+      '<feFlood flood-color="${colorToCssString(color)}" flood-opacity="1" result="flood">'
+      '</feFlood>' +
+      (swapLayers
+      ? '<feBlend in="SourceGraphic" in2="flood" mode="$feBlend"/>'
+      : '<feBlend in="flood" in2="SourceGraphic" mode="$feBlend"/>') +
+      '</filter></svg>';
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/bitmap_canvas.dart:1079]: These SVG snippets contain a lot of static content that need to be parsed from HTML. Wouldn't using the SVG API be more efficient?

File: lib/web_ui/test/golden_tests/engine/canvas_image_blend_mode_test.dart [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            Paint()
+              ..colorFilter = EngineColorFilter.mode(black, blendMode));
+      }
+      rc.restore();
+      await _checkScreenshot(rc, 'canvas_image_blend_group$blendGroup');
+    });
+  }
+}
+
+const String _flutterLogoBase64 =
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/test/golden_tests/engine/canvas_image_blend_mode_test.dart:109]: Done.

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:1079]: Didn't want to pull in svg library, not as big as dart:html but not small.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:1079]: Something tells me we already have it all included :)

It's OK. We can profile it later and see what the size/speed trade-off is like.

**Comment by ferhatb** [comment]
Mac Web Engine passed at https://ci.chromium.org/p/flutter/builders/try/Mac%20Web%20Engine/3029

**Pull request merged**